File: Problema2/challenge.java
/**
 * @file challenge.java
 * @author Alan Davide Bovo (alandavide.bovo.stud@ispascalcomandini.it)
 * @date 2024-01-03
 * Compile with: javac -g:none Problema2/challenge.java
 * Run with: java Problema2.challenge
 */
package Problema2;

// stringa decifrata: ReV3r53_1s_4w3s0m3
public class challenge {
    private final static String key = "*7^tVr4FZ#7S4RFNd2"; // chiave di cifratura
    private final static String encrypted = "xR\u0008G$G\u0007\u0019kPhgCa5~\t\u0001"; // testo cifrato
    
    public static boolean checkSignature(String signature){
        if(signature.length() != key.length()){
            return false; // se la lunghezza della chiave è diversa da quella della chiave di controllo
        }
        for(int i = 0; i < key.length(); i++){
            if((signature.charAt(i) ^ key.charAt(i)) != encrypted.charAt(i)){
                return false; // se il carattere della chiave è diverso da quello della chiave di controllo
            }
        }
        return true;
    }
    public static void main(String[] args){
        if(args.length != 1) {          // se il numero di argomenti è diverso da 1
            System.out.println("Usage: java Problema2.challenge <signature>");
            System.exit(1);
        }
        String signature = args[0];     // prendo la chiave

        if(checkSignature(signature))   // se la chiave è corretta
            System.out.println("Congrats! You have found the secret code!");
        else
            System.out.println("Sorry, the secret code is wrong!");
    }
}

File: Problema2/challenge.kt
/**
 * @file challenge.kt
 * @author Alan Davide Bovo (alandavide.bovo.stud@ispascalcomandini.it)
 * @date 2024-01-03
 * Compile with: kotlinc challenge.kt -include-runtime -d challenge.jar
 * Run with: java -jar challenge.jar
 */
import kotlin.io.*

const val key: String = "*7^tVr4FZ#7S4RFNd2"
const val encrypted: String = "xR\u0008G\$G\u0007\u0019kPhgCa5~\t\u0001"

fun checkSignature2(signature: String): Boolean {
    if(signature.length != key.length) { // se la lunghezza della firma è diversa dalla lunghezza della chiave
        return false
    }
    for(i in 0 until signature.length) { // per ogni carattere della firma
        // se il carattere della firma xor il carattere della chiave non è uguale al carattere cifrato
        if((signature[i].code xor key[i].code).toChar() != encrypted[i]) {
            return false
        }
    }
    return true
}

fun main(){
    print("Enter the secret code: ")        // chiede la firma
    val signature: String = readLine()!!    // legge la firma
    if(checkSignature2(signature)) {        // controlla la firma
        println("Congrats! You have found the secret code!")    // se è corretta, stampa il messaggio
    } else {
        println("Sorry, the secret code is wrong!")             // altrimenti stampa il messaggio
    }
}